### ISSUE TITLE
NullPointerException: no environment stored for lambda

### ISSUE THREAD

**Issue by XN137** [uber/NullAway#711]
hello, first of all thanks for this great tool

i was hitting the following error when running it over our codebase for the first time:

errorprone version: `2.17.0`
NullAway version: `0.10.7`
NullAway config: `-XepOpt:NullAway:AssertsEnabled=true -XepOpt:NullAway:ExhaustiveOverride=true -XepOpt:NullAway:CheckOptionalEmptiness=true -XepOpt:NullAway:HandleTestAssertionLibraries=true`
jdk version: `11`
java target version: `8`

code (modified for publication, but keeping the structure):
```
private void verifyCountZero(SomeStore<String, SomeData> store)
  verifyData(store, (count) -> Assert.assertEquals(0, (long) count));  // <-- line 106 (but different column)
}

private void verifyData(SomeStore<String, SomeData> store, Consumer<Long> assertFunction) {
  assertFunction.accept(StreamSupport.stream(store.findDataByCondition().spliterator(), true).count());
}
```

in the original code column 63 starts at:
```
 (long) count));
```

earlier in the file we have:
```
verifyData(store, (count) -> Assert.assertTrue(count > 0));
```
which causes no issues.

i can work around the problem with this lambda body:
```
Assert.assertEquals(Long.valueOf(0), count)
```
so i guess the cast is causing the problem somehow.

full error below:
```
[ERROR] /MyCode.java:[106,63] error: An unhandled exception was thrown by the Error Prone static analysis plugin.
[ERROR]      Please report this at https://github.com/google/error-prone/issues/new and include the following:
[ERROR]   
[ERROR]      error-prone version: 2.17.0
[ERROR]      BugPattern: NullAway
[ERROR]      Stack Trace:
[ERROR]      com.google.common.util.concurrent.UncheckedExecutionException: java.lang.NullPointerException: no environment stored for lambda
[ERROR]   	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2055)
[ERROR]   	at com.google.common.cache.LocalCache.get(LocalCache.java:3966)
[ERROR]   	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3989)
[ERROR]   	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4950)
[ERROR]   	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4956)
[ERROR]   	at com.uber.nullaway.dataflow.DataFlow.dataflow(DataFlow.java:157)
[ERROR]   	at com.uber.nullaway.dataflow.DataFlow.resultFor(DataFlow.java:290)
[ERROR]   	at com.uber.nullaway.dataflow.DataFlow.resultForExpr(DataFlow.java:266)
[ERROR]   	at com.uber.nullaway.dataflow.DataFlow.expressionDataflow(DataFlow.java:208)
[ERROR]   	at com.uber.nullaway.dataflow.AccessPathNullnessAnalysis.getNullness(AccessPathNullnessAnalysis.java:131)
[ERROR]   	at com.uber.nullaway.NullAway.nullnessFromDataflow(NullAway.java:2291)
[ERROR]   	at com.uber.nullaway.NullAway.mayBeNullExpr(NullAway.java:2254)
[ERROR]   	at com.uber.nullaway.NullAway.doUnboxingCheck(NullAway.java:1538)
[ERROR]   	at com.uber.nullaway.NullAway.matchTypeCast(NullAway.java:661)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.processMatchers(ErrorProneScanner.java:449)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitTypeCast(ErrorProneScanner.java:851)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitTypeCast(ErrorProneScanner.java:150)
[ERROR]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCTypeCast.accept(JCTree.java:2048)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.visitMethodInvocation(TreeScanner.java:509)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitMethodInvocation(ErrorProneScanner.java:751)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitMethodInvocation(ErrorProneScanner.java:150)
[ERROR]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1666)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.visitLambdaExpression(TreeScanner.java:559)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitLambdaExpression(ErrorProneScanner.java:702)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitLambdaExpression(ErrorProneScanner.java:150)
[ERROR]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCLambda.accept(JCTree.java:1831)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.visitMethodInvocation(TreeScanner.java:509)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitMethodInvocation(ErrorProneScanner.java:751)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitMethodInvocation(ErrorProneScanner.java:150)
[ERROR]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1666)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.visitExpressionStatement(TreeScanner.java:433)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitExpressionStatement(ErrorProneScanner.java:633)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitExpressionStatement(ErrorProneScanner.java:150)
[ERROR]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCExpressionStatement.accept(JCTree.java:1476)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:248)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:520)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:150)
[ERROR]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1048)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:206)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:740)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:150)
[ERROR]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:898)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:187)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:548)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:150)
[ERROR]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:144)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:560)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:150)
[ERROR]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:591)
[ERROR]   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
[ERROR]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:58)
[ERROR]   	at com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:43)
[ERROR]   	at com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:156)
[ERROR]   	at jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:132)
[ERROR]   	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1418)
[ERROR]   	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1365)
[ERROR]   	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:960)
[ERROR]   	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:311)
[ERROR]   	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:170)
[ERROR]   	at jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:75)
[ERROR]   	at jdk.internal.reflect.GeneratedMethodAccessor161.invoke(Unknown Source)
[ERROR]   	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR]   	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[ERROR]   	at org.codehaus.plexus.compiler.javac.JavacCompiler.compileInProcess0(JavacCompiler.java:694)
[ERROR]   	at org.codehaus.plexus.compiler.javac.JavacCompiler.compileInProcessWithProperClassloader(JavacCompiler.java:675)
[ERROR]   	at org.codehaus.plexus.compiler.javac.JavacCompiler.compileInProcess(JavacCompiler.java:664)
[ERROR]   	at org.codehaus.plexus.compiler.javac.JavacCompiler.performCompile(JavacCompiler.java:186)
[ERROR]   	at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:1209)
[ERROR]   	at org.apache.maven.plugin.compiler.TestCompilerMojo.execute(TestCompilerMojo.java:183)
[ERROR]   	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
[ERROR]   	at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2(MojoExecutor.java:370)
[ERROR]   	at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute(MojoExecutor.java:351)
[ERROR]   	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:215)
[ERROR]   	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:171)
[ERROR]   	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:163)
[ERROR]   	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
[ERROR]   	at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:210)
[ERROR]   	at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:195)
[ERROR]   	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[ERROR]   	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[ERROR]   	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[ERROR]   	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[ERROR]   	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[ERROR]   	at java.base/java.lang.Thread.run(Thread.java:829)
[ERROR]   Caused by: java.lang.NullPointerException: no environment stored for lambda
[ERROR]   	at java.base/java.util.Objects.requireNonNull(Objects.java:246)
[ERROR]   	at com.uber.nullaway.dataflow.CoreNullnessStoreInitializer.lambdaInitialStore(CoreNullnessStoreInitializer.java:86)
[ERROR]   	at com.uber.nullaway.dataflow.CoreNullnessStoreInitializer.getInitialStore(CoreNullnessStoreInitializer.java:41)
[ERROR]   	at com.uber.nullaway.dataflow.AccessPathNullnessPropagation.initialStore(AccessPathNullnessPropagation.java:203)
[ERROR]   	at com.uber.nullaway.dataflow.AccessPathNullnessPropagation.initialStore(AccessPathNullnessPropagation.java:135)
[ERROR]   	at org.checkerframework.nullaway.dataflow.analysis.ForwardAnalysisImpl.initInitialInputs(ForwardAnalysisImpl.java:342)
[ERROR]   	at org.checkerframework.nullaway.dataflow.analysis.AbstractAnalysis.init(AbstractAnalysis.java:376)
[ERROR]   	at org.checkerframework.nullaway.dataflow.analysis.ForwardAnalysisImpl.performAnalysis(ForwardAnalysisImpl.java:102)
[ERROR]   	at com.uber.nullaway.dataflow.DataFlow$1.load(DataFlow.java:93)
[ERROR]   	at com.uber.nullaway.dataflow.DataFlow$1.load(DataFlow.java:85)
[ERROR]   	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3533)
[ERROR]   	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2282)
[ERROR]   	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2159)
[ERROR]   	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2049)
[ERROR]   	... 122 more
[ERROR] -> [Help 1]
```

**Comment by msridhar** [comment]
Thanks for the report @XN137!  Unfortunately I'm not able to reproduce this crash yet.  Here is my attempt at a unit test based on what you wrote:

https://github.com/msridhar/NullAway/commit/d42122638a118fc95e3dd4c6317e8399d533a455

I don't see any crash for that test.  Do you see anything wrong with the test?  Like did I import a wrong library?  Can you check that commenting out the body of `verifyData()` in your code does not fix the crash?  Any further suggestions on how to repro would be helpful.  If you want to try yourself, you can check out the `[manu/issue-711](https://github.com/msridhar/NullAway/compare/manu/issue-711)` branch from my fork, modify the test I added, and then run `./gradlew :nullaway:test --tests NullAwayCoreTests.issue711` to see if the crash occurs.

One other thought.  Is the code you pasted above inside an anonymous class, or a local class declared within a method?  That might have something to do with it, though still not sure.

**Comment by XN137** [comment]
@msridhar thank you for the quick response.

i think i have identified the issue thanks to your example:
if we change
`"-XepOpt:NullAway:AnnotatedPackages=com.uber,com.ubercab,io.reactivex",`
to
`"-XepOpt:NullAway:AnnotatedPackages=com.whatever",`
then the test is hitting the same error:
![image](https://user-images.githubusercontent.com/1204398/211765601-69a8a94b-7bf4-4fde-9f82-88e30ab73f72.png)

so what i forgot to mention was that the error happens in a file of a java package, that is not actually part of the `AnnotatedPackages` list of packages and the above change reproduces that scenario.

you are right that the method body of `verifyData` doesnt matter.
similarly `SomeStore` and `SomeData` are not needed.
also the additional configuration options do not matter, so this is still reproducing:
```
  public void issue711() {
    makeTestHelperWithArgs(
        Arrays.asList(
            "-d",
            temporaryFolder.getRoot().getAbsolutePath(),
            "-XepOpt:NullAway:AnnotatedPackages=com.whatever"))
        .addSourceLines(
            "Test.java",
            "package com.uber;",
            "import java.util.function.Consumer;",
            "import org.junit.Assert;",
            "class Test {",
            "  private void verifyCountZero() {",
            "    verifyData((count) -> Assert.assertEquals(0, (long) count));",
            "  }",
            "  private void verifyData(Consumer<Long> assertFunction) {",
            "  }",
            "}")
        .doTest();
  }
```

**Comment by msridhar** [comment]
Thanks @XN137!  With that updated test I was able to easily figure out the problem.  #712 is up to fix

**Comment by msridhar** [comment]
FYI the fix for this issue is in version 0.10.8

**Comment by XN137** [comment]
@msridhar thanks for the hint, i can confirm that problem is fixed in the new version :+1:
